napi-inl.h

```diff
@@ -2498,12 +2498,61 @@ inline Error::Error(napi_env env, napi_value value) : ObjectReference(env, nullp
   if (value != nullptr) {
     napi_status status = napi_create_reference(env, value, 1, &_ref);
 
+    // Creates a wrapper object containg the error value (primitive types) and
+    // create a reference to this wrapper
+    if (status != napi_ok) {
+      napi_value wrappedErrorObj;
+      status = napi_create_object(env, &wrappedErrorObj);
+
+      NAPI_FATAL_IF_FAILED(status, "Error::Error", "napi_create_object");
+
+      status = napi_set_property(env,
+                                 wrappedErrorObj,
+                                 String::From(env, "errorVal"),
+                                 Value::From(env, value));
+      NAPI_FATAL_IF_FAILED(status, "Error::Error", "napi_set_property");
+
+      status = napi_set_property(env,
+                                 wrappedErrorObj,
+                                 String::From(env, "isWrapObject"),
+                                 Value::From(env, value));
+      NAPI_FATAL_IF_FAILED(status, "Error::Error", "napi_set_property");
+
+      status = napi_create_reference(env, wrappedErrorObj, 1, &_ref);
+    }
+
     // Avoid infinite recursion in the failure case.
     // Don't try to construct & throw another Error instance.
     NAPI_FATAL_IF_FAILED(status, "Error::Error", "napi_create_reference");
   }
 }
 
+inline Object Error::Value() const {
```

Comment:
I think it might make sense for this to return Value instead of Object (Values can be objects), the main concern is that would potentially be breaking. 

The reason is that we wrapping so that we can have a value versus just objects.  If its possible we might want two methods one which returns a Value and one which returns an Object in order to avoid breaking existing code.

Reply:
I tried setting the return type to Value and it seems to pass all the test suites.  Although I am not too sure if that breaks any conventions. (i.e, In the base Reference<T> class, ```Reference<T>::Value()``` will return an instance of type T, and Error is of type ```Object``` since it inherits ```ObjectReference```. Therefore people might expect Error::Value() to return an ```Object``` also) 

For the second part then, can I declare two public methods ```Napi::Value GetUnWrappedValue()``` and ```Object Value()``` ?

Reply:
If is possible to have 2 methods both called Value()  which return different types. If so that might make sense assuming the compiler will select the most appropriate one based on how you use the return value?

Reply:
Hmm, I am not too sure if it's possible to overload base on return value in C++ (without using templates) according to this post (https://stackoverflow.com/questions/9568852/overloading-by-return-type).

Reply:
After discussing in Node-API team meeting I think this is ok as  is.